File: frontend/web/src/lib/productContour.ts
/**
 * Адаптивный контур FinKlik — MVP: без общей системы налогообложения.
 * ООО/ЧУП/КФХ — только УСН + КУДиР; ИП — единый налог.
 */

export type LegalFormId = 'ip' | 'chup' | 'ooo' | 'kfh' | 'odo' | 'self_employed'

export type ProductContourId =
  | 'ip_single_tax'
  | 'org_usn'
  | 'lightweight'

export type ProductFeatures = {
  kudir: boolean
  chartOfAccounts: boolean
  fixedAssets: boolean
  fullLedger: boolean
  singleTax: boolean
  usn: boolean
  incomeTax: boolean
  employees: boolean
  counterparties: boolean
  documents: boolean
  reporting: boolean
}

export type ProductContour = {
  id: ProductContourId
  label: string
  accountingMode: 'simple' | 'advanced'
  features: ProductFeatures
}

const CONTOUR_FEATURES: Record<ProductContourId, ProductFeatures> = {
  ip_single_tax: {
    kudir: false,
    chartOfAccounts: false,
    fixedAssets: false,
    fullLedger: false,
    singleTax: true,
    usn: false,
    incomeTax: false,
    employees: true,
    counterparties: true,
    documents: true,
    reporting: true,
  },
  org_usn: {
    kudir: true,
    chartOfAccounts: false,
    fixedAssets: false,
    fullLedger: false,
    singleTax: false,
    usn: true,
    incomeTax: false,
    employees: true,
    counterparties: true,
    documents: true,
    reporting: true,
  },
  lightweight: {
    kudir: false,
    chartOfAccounts: false,
    fixedAssets: false,
    fullLedger: false,
    singleTax: false,
    usn: true,
    incomeTax: false,
    employees: false,
    counterparties: false,
    documents: true,
    reporting: true,
  },
}

const CONTOUR_META: Record<ProductContourId, { label: string; accountingMode: 'simple' | 'advanced' }> = {
  ip_single_tax: { label: 'ИП · единый налог', accountingMode: 'simple' },
  org_usn: { label: 'Организация · УСН и КУДиР', accountingMode: 'simple' },
  lightweight: { label: 'Упрощённый режим', accountingMode: 'simple' },
}

/** Допустимые tax_regime по ОПФ — MVP без общей системы. */
export const TAX_MODES_BY_LEGAL_FORM: Record<LegalFormId, { id: string; label: string }[]> = {
  ip: [{ id: 'single_tax', label: 'Единый налог' }],
  chup: [
    { id: 'usn_no_vat', label: 'УСН без НДС' },
    { id: 'usn_vat', label: 'УСН с НДС' },
  ],
  ooo: [
    { id: 'usn_no_vat', label: 'УСН без НДС' },
    { id: 'usn_vat', label: 'УСН с НДС' },
  ],
  kfh: [
    { id: 'usn_no_vat', label: 'УСН без НДС' },
    { id: 'usn_vat', label: 'УСН с НДС' },
  ],
  odo: [
    { id: 'usn_no_vat', label: 'УСН без НДС' },
    { id: 'usn_vat', label: 'УСН с НДС' },
  ],
  self_employed: [{ id: 'usn_no_vat', label: 'Упрощённый учёт' }],
}

export function normalizeLegalForm(raw: string | null | undefined): LegalFormId {
  const v = (raw || 'ip').trim().toLowerCase()
  if (v in TAX_MODES_BY_LEGAL_FORM) return v as LegalFormId
  return 'ip'
}

export function taxModesForLegalForm(form: string): { id: string; label: string }[] {
  return TAX_MODES_BY_LEGAL_FORM[normalizeLegalForm(form)]
}

export function isTaxRegimeValidForForm(legalForm: string, taxRegime: string): boolean {
  const modes = taxModesForLegalForm(legalForm)
  const tax = (taxRegime || '').trim().toLowerCase()
  if (modes.some((m) => m.id === tax)) return true
  // Legacy: ИП usn_no_vat / osn_vat → единый налог; организации osn_vat → УСН
  const form = normalizeLegalForm(legalForm)
  if (form === 'ip' && (tax === 'usn_no_vat' || tax === 'osn_vat')) return true
  if (form !== 'ip' && form !== 'self_employed' && tax === 'osn_vat') return true
  return false
}

export function resolveProductContour(
  legalForm: string | null | undefined,
  taxRegime: string | null | undefined,
): ProductContour {
  const form = normalizeLegalForm(legalForm)
  const tax = (taxRegime || 'usn_no_vat').trim().toLowerCase()

  if (form === 'self_employed') {
    const id: ProductContourId = 'lightweight'
    return { id, ...CONTOUR_META[id], features: CONTOUR_FEATURES[id] }
  }

  if (form === 'ip') {
    const id: ProductContourId = 'ip_single_tax'
    return { id, ...CONTOUR_META[id], features: CONTOUR_FEATURES[id] }
  }

  const id: ProductContourId = 'org_usn'
  return { id, ...CONTOUR_META[id], features: CONTOUR_FEATURES[id] }
}

export function hiddenRoutesForContour(contour: ProductContour): Set<string> {
  const hidden = new Set<string>()
  const f = contour.features

  if (!f.kudir) hidden.add('/accounting/kudir')
  if (!f.chartOfAccounts) hidden.add('/accounting/chart')
  if (!f.fixedAssets) hidden.add('/accounting/fixed-assets')
  if (!f.employees) {
    hidden.add('/employees')
    hidden.add('/employees/list')
  }
  if (!f.counterparties) hidden.add('/counterparties')

  return hidden
}

export function filterRoutes<T extends { to: string }>(items: T[], contour: ProductContour): T[] {
  const hidden = hiddenRoutesForContour(contour)
  return items.filter((i) => !hidden.has(i.to))
}

File: backend/api-gateway/app/services/product_contour.py
"""Адаптивный контур продукта — MVP без общей системы налогообложения."""

from __future__ import annotations

from typing import TypedDict

LEGAL_FORMS = frozenset({"ip", "chup", "ooo", "kfh", "odo", "self_employed"})

# MVP: osn_vat оставлен только для legacy-записей в БД (маппится в org_usn / single_tax)
TAX_MODES_BY_FORM: dict[str, frozenset[str]] = {
    "ip": frozenset({"single_tax", "usn_no_vat", "osn_vat"}),
    "chup": frozenset({"usn_no_vat", "usn_vat", "osn_vat"}),
    "ooo": frozenset({"usn_no_vat", "usn_vat", "osn_vat"}),
    "kfh": frozenset({"usn_no_vat", "usn_vat", "osn_vat"}),
    "odo": frozenset({"usn_no_vat", "usn_vat", "osn_vat"}),
    "self_employed": frozenset({"usn_no_vat"}),
}

TAX_REGIME_LABELS = {
    "usn_no_vat": "УСН без НДС",
    "usn_vat": "УСН с НДС",
    "osn_vat": "УСН (ранее общая система)",
    "single_tax": "Единый налог",
}

LEGAL_FORM_LABELS = {
    "ip": "ИП",
    "ooo": "ООО",
    "odo": "ОДО",
    "chup": "ЧУП",
    "kfh": "КФХ",
    "self_employed": "Самозанятый",
}


class ProductFeatures(TypedDict):
    kudir: bool
    chart_of_accounts: bool
    fixed_assets: bool
    full_ledger: bool
    single_tax: bool
    usn: bool
    income_tax: bool
    employees: bool
    counterparties: bool
    documents: bool
    reporting: bool


_CONTOUR_FEATURES: dict[str, ProductFeatures] = {
    "ip_single_tax": {
        "kudir": False,
        "chart_of_accounts": False,
        "fixed_assets": False,
        "full_ledger": False,
        "single_tax": True,
        "usn": False,
        "income_tax": False,
        "employees": True,
        "counterparties": True,
        "documents": True,
        "reporting": True,
    },
    "org_usn": {
        "kudir": True,
        "chart_of_accounts": False,
        "fixed_assets": False,
        "full_ledger": False,
        "single_tax": False,
        "usn": True,
        "income_tax": False,
        "employees": True,
        "counterparties": True,
        "documents": True,
        "reporting": True,
    },
    "lightweight": {
        "kudir": False,
        "chart_of_accounts": False,
        "fixed_assets": False,
        "full_ledger": False,
        "single_tax": False,
        "usn": True,
        "income_tax": False,
        "employees": False,
        "counterparties": False,
        "documents": True,
        "reporting": True,
    },
}


def normalize_legal_form(value: str | None) -> str:
    v = (value or "ip").strip().lower()
    return v if v in LEGAL_FORMS else "ip"


def is_tax_regime_valid(legal_form: str, tax_regime: str) -> bool:
    form = normalize_legal_form(legal_form)
    tax = (tax_regime or "").strip().lower()
    allowed = TAX_MODES_BY_FORM.get(form, TAX_MODES_BY_FORM["ip"])
    return tax in allowed


def resolve_product_contour(legal_form: str | None, tax_regime: str | None) -> dict:
    form = normalize_legal_form(legal_form)

    if form == "self_employed":
        contour_id = "lightweight"
        accounting_mode = "simple"
    elif form == "ip":
        contour_id = "ip_single_tax"
        accounting_mode = "simple"
    else:
        contour_id = "org_usn"
        accounting_mode = "simple"

    return {
        "id": contour_id,
        "accounting_mode": accounting_mode,
        "features": _CONTOUR_FEATURES[contour_id],
    }


def suggested_accounting_mode(legal_form: str | None, tax_regime: str | None) -> str:
    return resolve_product_contour(legal_form, tax_regime)["accounting_mode"]
